feat(gradle): allow specifying Gradle daemon heap size

snazy opened the pull request:
## Changes

This change introduces a heap size limitation for the Gradle daemon used
when updating the Gradle wrapper, defaults to 512m.

The heap size is limited to the globally configured limit, which
defaults to 512m. It can be configured on the repository level, but not
exceed the globally configured limit. A hard coded mimiumum heap size of
512m is enforced.

The configuration option is named `toolSettings`, which has two
properties for the maximum and initial heap size. See
`docs/usage/configuration-options#toolSettings` for more details.

## Context

Please select one of the following:

- [X] This closes an existing Issue, Closes: #39558
- [ ] This doesn't close an Issue, but I accept the risk that this PR may be closed if maintainers disagree with its opening or implementation

## AI assistance disclosure

<!-- We request this information to assist reviewers in identifying AI-generated errors and other issues specific to AI usage. While we typically permit the use of AI tools, we appreciate being notified when they are employed. -->

Did you use AI tools to create any part of this pull request?

Please select one option and, if yes, briefly describe how AI was used (e.g., code, tests, docs) and which tool(s) you used.

- [X] No — I did not use AI for this contribution.
- [ ] Yes — minimal assistance (e.g., IDE autocomplete, small code completions, grammar fixes).
- [ ] Yes — substantive assistance (AI-generated non‑trivial portions of code, tests, or documentation).
- [ ] Yes — other (please describe):

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [X] Code inspection only, or
- [X] Newly added/modified unit tests, or
- [ ] No unit tests, but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

The public repository: <URL>

<!-- If you have any suggestions about this PR template, edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. You can commit as many times as you need in this branch. -->

jamietanna reviewed (changes requested):
Thanks! Couple of things:

- Adding the `--stop` and `--no-daemon` would be great as separate changes (i.e. new PR), so they can be released + tested independently, as they should be safe to handle on their own
- Via #39558 we should make this configurable, at least for Global config, so it's not defaulted (and i.e. to allow Mend Renovate Enterprise Cloud users to have increased memory allocated, if they wish)
- In the future, we can look at adding repo-level config to tune this

Thanks!

RahulGautamSingh reviewed (commented):
+1 for making it configurable (at global level)

jamietanna reviewed (commented):
Looks on the right tracks - mind addressing the `bigint`/`number` issue?

snazy commented:
Okay, I think, I have something that's somewhat okay.
I've removed everything from this PR that's not strictly related to the memory issue during wrapper upgrades issue.

The config is added to `RepoGlobalConfig`, but I'm not sure whether that's the right place. It feels a bit odd to add the gradle-wrapper manager-specific config there, although there seem to be other specific settings as well.
Would `ManagerConfig` be a better place, considering the added config can be set to different values (incl defaults) for the `gradle` manager?

viceice commented:
conflicted

viceice commented on lib/config/global.ts:
```diff
@@ -33,6 +33,7 @@ export class GlobalConfig {
     'exposeAllEnv',
     'gitTimeout',
     'githubTokenWarn',
+    'gradleWrapper',
```
not enough , need a config option. check other global options for samples

snazy commented:
Hm, seems I'm running into some "configuration-options deadlock":
`test/docs/documentation.spec.ts` complains about `docs/usage/self-hosted-configuration.md` the unexpected `jvmMaxMemory` and `jvmMemory` options (those options are declared with `globalOnly: true`).
Without `globalOnly: true`, `pnpm create-json-schema` fails since #38619 though.
I'm a bit lost on how to solve this.

viceice commented on lib/config/options/index.ts:
```diff
@@ -3344,6 +3344,35 @@ const options: Readonly<RenovateOptions>[] = [
+    default: {
+      jvmMaxMemory: 256,
+      jvmMemory: 256,
+    },
```
```suggestion
    },
    freeChoice: true,
```

don't validate children, just document supported values

viceice reviewed (commented):
my idea for workaround

snazy commented on lib/config/options/index.ts:
```diff
@@ -3344,6 +3344,35 @@ const options: Readonly<RenovateOptions>[] = [
+    parents: ['gradleWrapper'],
+    stage: 'global',
+    globalOnly: true,
+  },
```
The suggestions let `pnpm check` pass :)

snazy commented on lib/config/types.ts:
```diff
@@ -803,3 +805,8 @@ export interface BumpVersionConfig {
+
+export interface GradleWrapperOptions {
+  jvmMaxMemory?: bigint;
+  jvmMemory?: bigint;
```
Yea, I wanted it to be an integer, not a float - so I opted for `bigint` over `number`.
But I think I found a better way, as in a new commit.

jamietanna reviewed (changes requested):
I've got some changes I'll push into the branch - please hold off on any force pushes

viceice commented:
no squash needed, we'll squash on merge anyways

snazy committed:
feat(gradle): allow specifying Gradle daemon heap size

This change introduces a heap size limitation for the Gradle daemon used
when updating the Gradle wrapper, defaults to 512m.

The heap size is limited to the globally configured limit, which
defaults to 512m. It can be configured on the repository level, but not
exceed the globally configured limit. A hard coded mimiumum heap size of
512m is enforced.

The configuration option is named `toolSettings`, which has two
properties for the maximum and initial heap size. See
`docs/usage/configuration-options#toolSettings` for more details.

snazy commented:
@jamietanna and nice to have the repo-level setting :)

jamietanna commented on lib/config/types.ts:
```diff
@@ -607,6 +610,12 @@ export interface RenovateOptionBase {
+   * and docs/usage/self-hosted-configuration.md.
+   * This is purely a documentation validation option, no runtime implications.
+   */
+  docAlsoGlobal?: boolean;
```
Can this change (and in `test/docs/documentation.spec.ts`) please be a separate PR?

snazy commented:
:tada: `codecov/patch — 100.00% of diff hit`

jamietanna reviewed (commented):
One comment, and an ask to test against a repo, otherwise LGTM

Given a bit of code rewriting from me, not sure if Michael you want to re-review to confirm you're happy?

snazy commented on docs/usage/configuration-options.md:
````diff
@@ -1890,6 +1890,34 @@ If you need to modify this path, for example in order to ignore directories, you
 }
 ```
 
+## gradleWrapper
````
Hm, looks like the property name `manager` conflicts with other settings.
Wonder whether `manager` could become something like a type to define limits/constraints for "all the called tools", like this:
```json
"tools": {
  "jvmMaxMemory": 512,

  ... potential future settings
  'goMaxMemory": .........
}
```
WDYT?

snazy replied:
Pushed a commit using `tools`

jamietanna reviewed (commented):
Re 512 as the default, LGTM - it's also what Gradle defaults to https://docs.gradle.org/9.3.1/userguide/config_gradle.html

jamietanna reviewed (dismissed):
I'll let you rebase if you wanted to clean up the history (or if not, LMK and we'll squash-merge it as we usually do) and rewrite commits/add any additional context

jamietanna commented:
Yep, exactly - my plan is that 2026-02-09, we'll start testing on a subset of repos, and either 2026-02-12 or 2026-02-16 this will land for all users

snazy commented:
Feel free to add projectnessie/nessie to those repos ;)
(I'm a maintainer of that one.)

jamietanna commented:
@snazy you're currently running [43.6.0](https://github.com/renovatebot/renovate/releases/tag/43.6.0) (2026-02-10) and we'll be continuing to update it over the week - LMK if you spot anything not working!
